File: program_topic_go/others/order/merge_sort.go
package main

func merge_sort(arr []int) {
	l := len(arr)
	if l == 0 || arr == nil {
		return
	}

	merge_helper(arr, 0, l-1)
}

func merge_helper(arr []int, low, high int) {
	mid := (low + high) >> 1
	if low < high {
		merge_helper(arr, low, mid)
		merge_helper(arr, mid+1, high)
		merge_core(arr, low, mid, high)
	}
}

func merge_core(arr []int, left, mid, right int) {
	tmp := make([]int, right-left+1)
	i, j, k := left, mid+1, 0
	for i <= mid && j <= right {
		if arr[i] < arr[j] {
			tmp[k] = arr[i]
			i++
		} else {
			tmp[k] = arr[j]
			j++
		}
		k++
	}

	for i <= mid {
		tmp[k] = arr[i]
		k++
		i++
	}

	for j <= right {
		tmp[k] = arr[j]
		k++
		j++
	}

	for m := 0; m < len(tmp); m++ {
		arr[left+m] = tmp[m]
	}
}

func main() {
	arr := []int{7, 0, 1, 9, 4, 6, 3, 5, 2, 8}
	print_arr(arr)
	merge_sort(arr)
	print_arr(arr)
}

//func print_arr(arr []int) {
//	l := len(arr)
//	if arr == nil || l == 0 {
//		return
//	}
//
//	for i := 0; i < l-1; i++ {
//		fmt.Print(arr[i], "\t")
//	}
//
//	fmt.Println(arr[l-1])
//}

File: program_topic_java/src/others/order/MergeSort.java
package others.order;

import java.lang.reflect.Method;

//归并排序
public class MergeSort {
    public static void main(String[] args) {
        MergeSort mergeSort = new MergeSort();
        int[] arr = {8, 0, 5, 7, 9, 1, 4, 2, 6, 3};
        mergeSort.printArray(arr);
        mergeSort.mergeSort(arr);
        mergeSort.printArray(arr);
    }

    public void mergeSort(int[] arr) {
        int size = arr.length;
        if (size == 0 || arr == null) {
            return;
        }

        mergeHelper(arr, 0, size - 1);
    }

    public void mergeHelper(int[] arr, int low, int high) {
        int mid = (low + high) >> 1;
        if (low < high) {
            mergeHelper(arr, low, mid);
            mergeHelper(arr, mid + 1, high);
            mergeCore(arr, low, mid, high);
        }
    }

    public void mergeCore(int[] arr, int left, int mid, int right) {
        int[] tmp = new int[right - left + 1];
        int i = left;
        int j = mid + 1;
        int k = 0;
        while (i <= mid && j <= right) {
            if (arr[i] < arr[j]) {
                tmp[k++] = arr[i++];
            } else {
                tmp[k++] = arr[j++];
            }
        }

        while (i <= mid) {
            tmp[k++] = arr[i++];
        }

        while (j <= right) {
            tmp[k++] = arr[j++];
        }

        for (int m = 0; m < tmp.length; m++) {
            arr[left + m] = tmp[m];
        }
    }

    public void printArray(int[] arr) {
        int size = arr.length;
        if (size == 0 || arr == null) {
            return;
        }

        for (int i = 0; i < size - 1; i++) {
            System.out.print(arr[i] + "\t");
        }
        System.out.println(arr[size - 1]);
    }
}
